block.go

```diff
@@ -0,0 +1,272 @@
+		}).
+		ReadSliceOfArraysOf32Bytes(&m.Parents, deSeriMode, serializer.SeriLengthPrefixTypeAsByte, &blockParentArrayRules, func(err error) error {
+			return fmt.Errorf("unable to deserialize block parents: %w", err)
+		}).
+		ReadPayload(&m.Payload, deSeriMode, deSeriCtx, blockPayloadGuard.ReadGuard, func(err error) error {
+			return fmt.Errorf("unable to deserialize block's inner payload: %w", err)
+		}).
+		ReadNum(&m.Nonce, func(err error) error {
+			return fmt.Errorf("unable to deserialize block nonce: %w", err)
+		}).
+		ConsumedAll(func(leftOver int, err error) error {
+			return fmt.Errorf("%w: unable to deserialize block: %d bytes are still available", err, leftOver)
+		}).
+		Done()
+}
+
+func (m *Block) Serialize(deSeriMode serializer.DeSerializationMode, deSeriCtx interface{}) ([]byte, error) {
+	data, err := serializer.NewSerializer().
+		Do(func() {
+			if deSeriMode.HasMode(serializer.DeSeriModePerformLexicalOrdering) {
+				m.Parents = serializer.RemoveDupsAndSortByLexicalOrderArrayOf32Bytes(m.Parents)
+			}
+		}).
+		WriteNum(m.ProtocolVersion, func(err error) error {
+			return fmt.Errorf("unable to serialize block protocol version: %w", err)
+		}).
+		Write32BytesArraySlice(m.Parents, deSeriMode, serializer.SeriLengthPrefixTypeAsByte, &blockParentArrayRules, func(err error) error {
+			return fmt.Errorf("unable to serialize block parents: %w", err)
+		}).
+		WritePayload(m.Payload, deSeriMode, deSeriCtx, blockPayloadGuard.WriteGuard, func(err error) error {
+			return fmt.Errorf("unable to serialize block inner payload: %w", err)
+		}).
+		WriteNum(m.Nonce, func(err error) error {
+			return fmt.Errorf("unable to serialize block nonce: %w", err)
+		}).
+		Serialize()
+	if err != nil {
+		return nil, err
+	}
+	if len(data) > BlockBinSerializedMaxSize {
+		return nil, fmt.Errorf("%w: size %d bytes", ErrBlockExceedsMaxSize, len(data))
+	}
+	return data, nil
+}
+
+func (m *Block) MarshalJSON() ([]byte, error) {
+	jBlock := &jsonBlock{
+		ProtocolVersion: int(m.ProtocolVersion),
+	}
+	jBlock.Parents = make([]string, len(m.Parents))
+	for i, parent := range m.Parents {
+		jBlock.Parents[i] = EncodeHex(parent[:])
+	}
+	jBlock.Nonce = EncodeUint64(m.Nonce)
+	if m.Payload != nil {
+		jsonPayload, err := m.Payload.MarshalJSON()
+		if err != nil {
+			return nil, err
+		}
+		rawMsgJsonPayload := json.RawMessage(jsonPayload)
+		jBlock.Payload = &rawMsgJsonPayload
+	}
+	return json.Marshal(jBlock)
+}
+
+func (m *Block) UnmarshalJSON(bytes []byte) error {
+	jBlock := &jsonBlock{}
+	if err := json.Unmarshal(bytes, jBlock); err != nil {
+		return err
+	}
+	seri, err := jBlock.ToSerializable()
+	if err != nil {
+		return err
+	}
+	*m = *seri.(*Block)
+	return nil
+}
+
+// jsonBlock defines the JSON representation of a Block.
+type jsonBlock struct {
+	ProtocolVersion int `json:"protocolVersion"`
+	// The hex encoded IDs of the referenced parent blocks.
+	Parents []string `json:"parentBlockIds"`
```

Comment:
do we maybe want to just call it parents in JSON too?

Reply:
sounds good to me

Reply:
Should we do the same for the API block metadata response ?

Reply:
I renamed them